File: Enose1-main/app/static/js/automation.js
/**
 * E-Nose — Automation Pipeline page
 */

const PIPELINE_STEPS = ['IDLE', 'PURGE', 'WAIT', 'SAMPLE', 'FLUSH'];

let pipelineRunning = false;

function updateStepUI(state, stepElapsed, stepTotal) {
  PIPELINE_STEPS.forEach((name, i) => {
    const el = document.getElementById(`step-${name}`);
    if (!el) return;
    el.classList.remove('active', 'done');
    const idx = PIPELINE_STEPS.indexOf(state);
    if (i < idx) el.classList.add('done');
    else if (i === idx && state !== 'IDLE') el.classList.add('active');
  });

  // Progress bar
  const bar = document.getElementById('pipeline-progress');
  if (bar && stepTotal > 0) {
    bar.style.width = `${Math.min(100, (stepElapsed / stepTotal) * 100)}%`;
  } else if (bar) {
    bar.style.width = '0%';
  }

  // Status label
  const lbl = document.getElementById('pipeline-state-label');
  if (lbl) lbl.textContent = state === 'IDLE' ? 'Chờ' : state;

  pipelineRunning = state !== 'IDLE';
  const btnStart = document.getElementById('btn-auto-start');
  const btnStop  = document.getElementById('btn-auto-stop');
  if (btnStart) btnStart.disabled = pipelineRunning;
  if (btnStop)  btnStop.disabled  = !pipelineRunning;
}

async function loadStatus() {
  try {
    const s = await API.get('/api/automation/status');
    updateStepUI(s.state, 0, 0);
  } catch (_) {}
}

async function loadProjects() {
  const projects = await API.get('/api/projects/');
  const sel = document.getElementById('auto-proj-select');
  if (!sel) return;
  sel.innerHTML = '<option value="">-- Chọn dự án --</option>';
  projects.forEach(p => sel.insertAdjacentHTML('beforeend', `<option value="${p.id}">${p.name}</option>`));
}

async function startPipeline() {
  const projectId = document.getElementById('auto-proj-select')?.value;
  if (!projectId) { showToast('Chọn dự án trước', 'warning'); return; }

  const cfg = {
    project_id:      parseInt(projectId),
    session_name:    document.getElementById('auto-sess-name')?.value || 'Auto Session',
    label:           document.getElementById('auto-sess-label')?.value || '',
    purge_duration:  parseInt(document.getElementById('auto-purge')?.value || 30),
    sample_delay:    parseInt(document.getElementById('auto-delay')?.value || 5),
    sample_duration: parseInt(document.getElementById('auto-sample')?.value || 60),
    flush_duration:  parseInt(document.getElementById('auto-flush')?.value || 20),
    pump_speed:      Math.round((parseInt(document.getElementById('auto-pump')?.value || 80) / 100) * 255),
  };

  try {
    const r = await API.post('/api/automation/start', cfg);
    if (r.ok) showToast('Pipeline đã khởi động', 'success');
    else showToast(r.message || 'Lỗi', 'error');
  } catch (e) { showToast('Lỗi kết nối', 'error'); }
}

async function stopPipeline() {
  try {
    await API.post('/api/automation/stop');
    showToast('Đã dừng pipeline', 'warning');
  } catch (e) { showToast('Lỗi', 'error'); }
}

document.addEventListener('DOMContentLoaded', () => {
  loadStatus();
  loadProjects();

  document.getElementById('btn-auto-start')?.addEventListener('click', startPipeline);
  document.getElementById('btn-auto-stop')?.addEventListener('click', stopPipeline);

  document.getElementById('auto-pump')?.addEventListener('input', e => {
    const el = document.getElementById('auto-pump-val');
    if (el) el.textContent = e.target.value;
  });

  document.addEventListener('pipeline_status', (e) => {
    const d = e.detail;
    updateStepUI(d.state, d.step_elapsed, d.step_total);
  });
});

File: Enose1-main/app/api/automation.py
from flask import Blueprint, request, jsonify
from app.automation import pipeline
from app.models import session as sess_model

bp = Blueprint('automation', __name__)


@bp.get('/status')
def status():
    eng = pipeline.get_engine()
    if not eng:
        return jsonify({'state': 'IDLE', 'is_running': False})
    return jsonify({
        'state': eng.state,
        'is_running': eng.is_running(),
        'session_id': eng.current_session_id,
        'config': eng.config,
    })


@bp.post('/start')
def start():
    data = request.json or {}
    eng = pipeline.get_engine()
    if not eng:
        return jsonify({'error': 'Engine not initialized'}), 500

    # Create or use existing session
    project_id = data.get('project_id')
    session_id = data.get('session_id')
    if not session_id and project_id:
        session_name = data.get('session_name', 'Auto Session')
        session_id = sess_model.create(project_id, session_name, data.get('label', ''))

    if not session_id:
        return jsonify({'error': 'session_id or project_id required'}), 400

    cfg = {
        'purge_duration':  data.get('purge_duration', 30),
        'sample_delay':    data.get('sample_delay', 5),
        'sample_duration': data.get('sample_duration', 60),
        'flush_duration':  data.get('flush_duration', 20),
        'pump_speed':      data.get('pump_speed', 200),
    }
    ok, msg = eng.start(session_id, cfg)
    return jsonify({'ok': ok, 'message': msg, 'session_id': session_id})


@bp.post('/stop')
def stop():
    eng = pipeline.get_engine()
    if eng:
        eng.stop()
    return jsonify({'ok': True})
